File: thirvusoft/thirvusoft_customizations/report/ts_scrum_tool/ts_scrum_tool.py
# Copyright (c) 2022, ThirvuSoft Private Limited and contributors
# For license information, please see license.txt


import frappe
from frappe import _
from frappe.utils import nowdate
def execute(filters=None):
	from_date = filters.get("from_date")
	to_date = filters.get("to_date")
	employee = filters.get("employee")
	status = filters.get("status")
	conditions = ""
	if from_date or to_date or employee or status:
		conditions = " where 1 = 1 "
		if from_date and to_date:
			conditions += " and task.exp_start_date between '{0}' and '{1}' ".format(from_date, to_date)
		if employee:
			conditions += " and task.ts_assigned_team_member ='{0}' ".format(employee)
		if status:
			conditions += " and task.status = '{0}' ".format(status)
	report_data = frappe.db.sql("""select task.name,task.ts_assigned_team_member,emp.employee_name,task.exp_start_date,task.expected_time,task.ts_assigned_ci_name,actual_time,task.status
									from tabTask as task
									left outer join tabEmployee as emp on
  										  task.ts_assigned_team_member = emp.name
									{0}
								""".format(conditions))
	data = [list(i) for i in report_data]

	columns = get_columns()
	return columns, data

def get_columns():
	columns = [
		_("Task ID") + ":Link/Task:150",
		_("Employee") + ":Data:200",
		{
			"fieldname":"employee_name",
			"fieldtype": "Data",
			"hidden": 1,
			
		},
		_("Expected Start Date") + ":Data:180",
		_("Expected time") + ":data:130",
		_("Assigned PM") + ":Data:180",
		_("Actual Time") + ":data:100",
		_("Task Status") + ":Data:180",

		]
	
	return columns

@frappe.whitelist()
def task_status(task_name,task_status):
    task_doc= frappe.get_doc('Task',task_name)
    if (task_status == "Completed"):
        task_doc.update({
            'status':task_status,
            'completed_on':nowdate()
        })
        task_doc.save()
        frappe.db.commit()
    else:
        task_doc.update({
                'status':task_status
            })
        task_doc.save()
        frappe.db.commit()
		

File: thirvusoft/thirvusoft_customizations/report/ts_scrum_tool/ts_scrum_tool.js
// Copyright (c) 2022, ThirvuSoft Private Limited and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["TS Scrum Tool"] = {
	"filters": [
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"width": "80"
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"width": "80"
		},
		{
			"fieldname":"status",
			"label": __("Task Status"),
			"fieldtype": "Select",
			"options": ["","Pending Review","Hold","Completed","Cancelled"],
			"width": "100"
		},
		{
			"fieldname":"employee",
			"label": __("Employee"),
			"fieldtype": "Link",
			"options": "Employee",
			"width": "100"
		}

	],
	
	get_datatable_options(options) {
        return Object.assign(options, {
            checkboxColumn: true
        });
    },
	onload: function() {
		frappe.query_report.page.add_inner_button(__("Update Status"), function() {
				var d = new frappe.ui.Dialog({
					size:"large",
					title: "Action",
					fields: [
						{label:'Task Status',fieldname:'task_status',fieldtype:'Select',options: ["Hold","Open","Working","PR Opened","PR Merged","PR Conflicts","PR Closed","Deployed to Production","Development Testing Started","Development Testing Ended","CI Verified","Client Satisfied","Overdue","Completed","Cancelled"],reqd:1},
						{fieldtype:'Column Break'},
					],
					primary_action:function(data){
						var status = data.task_status
						change_status(status)
						d.hide()
						frappe.msgprint("Status Updated");
					}
					
						})
						d.show()	
		});
		
	}
};

var change_status = function(status){	
		let checked_rows_indexes = frappe.query_report.datatable.rowmanager.getCheckedRows();
		let checked_rows = checked_rows_indexes.map(i => frappe.query_report.data[i]);
			var task_list = []
			for(var i=0;i<checked_rows.length;i++)
				{
					task_list.push(checked_rows[i].task_id)
					
				}
				console.log(task_list)
				for(let i in task_list){
						frappe.call({
							method: "thirvusoft.thirvusoft_customizations.custom.python.reports.task_status",
							args: {
								name: task_list[i],
								value: status
							},
						});
				}	
}



